### PR TITLE
Build Commerce APIs with Rest Builder

### PR DESCRIPTION
**Pull request by ZoltanTakacs**

### CODE REVIEW

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+	@PUT
+	@Path("/taxCategory/{id}")
+	@Produces(
+		{
+			"application/json", "application/json", "application/json",
+			"application/xml", "application/xml", "application/xml"
+		}
+	)
+	@Tags(value = {@Tag(name = "TaxCategory")})
+	public Error updateTaxCategory(
```
Comment (ZoltanTakacs) [inline headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java]: - Why do we returning the Error object here?

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+	@Override
+	@GET
+	@Path("/commerceAdminSiteSetting/{groupId}/taxCategory/")
+	@Produces(
+		{
+			"application/json", "application/json", "application/json",
+			"application/json", "application/xml", "application/xml",
+			"application/xml", "application/xml"
```
Comment (ZoltanTakacs) [inline headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java]: - There are duplicated media types here. Can we trim the list and make sure we only have unique ones?

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+@Generated("")
+@Path("/v1.0")
+public abstract class BaseTaxCategoryResourceImpl
+	implements TaxCategoryResource {
+
+	@Override
+	@DELETE
+	@Path("/taxCategory/{id}")
+	@Produces({"application/json", "application/xml"})
+	@Tags(value = {@Tag(name = "TaxCategory")})
```
Comment (ZoltanTakacs) [inline headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java]: - What is the tag used for? Can you please give us some information on it?

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+			"application/json", "application/json", "application/json",
+			"application/json", "application/xml", "application/xml",
+			"application/xml", "application/xml"
+		}
+	)
+	@Tags(value = {@Tag(name = "TaxCategory")})
+	public TaxCategory upsertTaxCategory(
+			@PathParam("groupId") Long groupId, TaxCategory taxCategory,
+			TaxCategory taxCategory)
+		throws Exception {
```
Comment (ZoltanTakacs) [inline headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java]: We have 2 `taxCategory` parameter in this method signature and I saw this issue in `BaseMeasurementUnitResourceImpl` and `BaseMeasurementUnitResourceTestCase` as well. Can you check this?

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/jaxrs/application/HeadlessCommerceAdminSiteSettingApplication.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * @author Zoltán Takács
+ * @generated
+ */
+@Component(
+	property = {
+		"osgi.jaxrs.application.base=/headless-commerce-admin-site-setting",
+		"osgi.jaxrs.extension.select=(osgi.jaxrs.name=Liferay.Vulcan)",
```
Comment (ZoltanTakacs) [inline headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/jaxrs/application/HeadlessCommerceAdminSiteSettingApplication.java]: - What is the purpose of this extension? What does it provide?

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java (hunk 1 above)
Comment (petershin) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java]: Sent a patch to update the return type to `Response` when multiple `responses` are defined in the `rest-openapi.yaml` file

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java (hunk 2 above)
Comment (petershin) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java]: https://github.com/brianchandotcom/liferay-portal/pull/69411

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java (hunk 3 above)
Comment (petershin) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java]: The tag is used for grouping paths in the [swagger editor](https://editor.swagger.io/). Pablo Agulla requested the feature and can provide more info

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java (hunk 4 above)
Comment (petershin) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java]: When the `requestBody` `mediaType` is specified, the generator will automatically append a `dto` parameter to the method (i.e. the second `taxCategory`).

Offhand, I think the `rest-openapi.yaml` can be updated to fix the issue.  I'll look into this more and get back to you

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/jaxrs/application/HeadlessCommerceAdminSiteSettingApplication.java (hunk 5 above)
Comment (petershin) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/jaxrs/application/HeadlessCommerceAdminSiteSettingApplication.java]: The extension provides several features, here's a few:
```
- exception mappers to translate errors to the right status code
- readers/writers to parse JSON and understand Multipart
- providers for pagination and the ability filter fields
```

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java (hunk 1 above)
Comment (ZoltanTakacs) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java]: @petershin Unfortunately it is not so easy :/ We have defined multiple response types for the users to document what kind of error they may encounter on the request.

So for now, as of this change, all of the generated methods returns just a `Response` for us. 

- The Commerce Generator was generating the return type based on the first Success Type Family status. -> https://github.com/ZoltanTakacs/com-liferay-commerce/blob/7.1.x-next/commerce-openapi-util/src/main/java/com/liferay/commerce/openapi/util/Response.java#L45

See https://docs.oracle.com/javaee/7/api/javax/ws/rs/core/Response.Status.Family.html#SUCCESSFUL

The other statuses are most probably Errors which were handled by ExceptionMappers which will eventually turn those Exception to Error responses with the appropriate HTTP status.

----

- Can we change this logic and return with the type defined by the `Response.Status.Family.html#SUCCESSFUL` family's lowest code number? 

Simple example would be: In the Specification:
- have status with 200 -> Type1
- have status with 204 with no content

then the generated return type would be `Type1`.

/cc @igorbeslic

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java (hunk 4 above)
Comment (ZoltanTakacs) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java]: @petershin I don't really see how we can update the spec other than removing one of the supported media type, but it not really an option 🤔 
- I think we should generate two different methods if the schemas are different and use one dto in the 1st and the other dto in the 2nd method signature.
- But usually these schemas are the same, _(it's just used by the client to set the proper content type)_ so we can generate 1 method signature with only including the schema once.

/cc @igorbeslic

File: headless/headless-commerce/headless-commerce-admin-api/src/main/java/com/liferay/headless/commerce/admin/resource/v2_0/AccountResource.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+	public Response putMediaType1AccountId(String id, Account account)
+		throws Exception;
+
+	public Response putMediaType2AccountId(String id, Account account)
+		throws Exception;
+
+	public Response getAddresses(String id) throws Exception;
+
+	public Response postMediaType1AccountIdAddres(String id, Address address)
```
Comment (ZoltanTakacs) [inline headless/headless-commerce/headless-commerce-admin-api/src/main/java/com/liferay/headless/commerce/admin/resource/v2_0/AccountResource.java]: @petershin with the latest version of generator, it generates method names like this: `postMediaType1AccountIdAddres` see the other below :/ 
- Can you check this issue? Nothing has changed in the spec.

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/jaxrs/application/HeadlessCommerceAdminSiteSettingApplication.java (hunk 5 above)
Comment (ZoltanTakacs) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/jaxrs/application/HeadlessCommerceAdminSiteSettingApplication.java]: @petershin Who is the responsible person for maintaining that module (`portal-vulcan`)? I want to get more info about the filtering feature.. Is it documented somewhere?

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java (hunk 1 above)
Comment (petershin) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java]: Taking a look at this today

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java (hunk 4 above)
Comment (petershin) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/resource/v1_0/BaseTaxCategoryResourceImpl.java]: I don't see this issue with the latest generated code.  Will send an updated pull to you.

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/jaxrs/application/HeadlessCommerceAdminSiteSettingApplication.java (hunk 5 above)
Comment (petershin) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/jaxrs/application/HeadlessCommerceAdminSiteSettingApplication.java]: Javier Gamarra will probably be able to provide the information you need

File: headless/headless-commerce/headless-commerce-admin-api/src/main/java/com/liferay/headless/commerce/admin/resource/v2_0/AccountResource.java (hunk 6 above)
Comment (petershin) [reply headless/headless-commerce/headless-commerce-admin-api/src/main/java/com/liferay/headless/commerce/admin/resource/v2_0/AccountResource.java]: Javier Gamarra requested this change and it was added in yesterday.  He'll be able to provide more details on why it was needed.

**Comment by petershin** [comment]
Sent to Zoltan: https://github.com/ZoltanTakacs/com-liferay-commerce/pull/103

File: headless/headless-commerce/headless-commerce-admin-api/src/main/java/com/liferay/headless/commerce/admin/resource/v2_0/AccountResource.java (hunk 6 above)
Comment (ZoltanTakacs) [reply headless/headless-commerce/headless-commerce-admin-api/src/main/java/com/liferay/headless/commerce/admin/resource/v2_0/AccountResource.java]: Hi @nhpatt can you please explain this change?

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/jaxrs/application/HeadlessCommerceAdminSiteSettingApplication.java (hunk 5 above)
Comment (ZoltanTakacs) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/jaxrs/application/HeadlessCommerceAdminSiteSettingApplication.java]: @nhpatt I didn't check the impl for filtering, can you update us how it's working currently? Thanks in advance.

File: headless/headless-commerce/headless-commerce-admin-api/src/main/java/com/liferay/headless/commerce/admin/resource/v2_0/AccountResource.java (hunk 6 above)
Comment (nhpatt) [reply headless/headless-commerce/headless-commerce-admin-api/src/main/java/com/liferay/headless/commerce/admin/resource/v2_0/AccountResource.java]: mmm... yesterday I asked to generate 2 methods if there are two request bodies (one multipart and one json, for example)

Comment (nhpatt) [reply headless/headless-commerce/headless-commerce-admin-api/src/main/java/com/liferay/headless/commerce/admin/resource/v2_0/AccountResource.java]: We can rollback this change and I can send what I was planning later if you want :) 

In your case it should generate only 1 method, I was thinking on this use case:

content:
                    application/json:
                        schema:
                            $ref: "#/components/schemas/UserAccount"
                    multipart/form-data:
                        schema:
                            properties:
                                file:
                                    format: binary
                                    type: string
                                userAccount:
                                    $ref: "#/components/schemas/UserAccount"
                            type: object

File: headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/jaxrs/application/HeadlessCommerceAdminSiteSettingApplication.java (hunk 5 above)
Comment (nhpatt) [reply headless/headless-commerce/headless-commerce-admin-site-setting-impl/src/main/java/com/liferay/headless/commerce/admin/site/setting/internal/jaxrs/application/HeadlessCommerceAdminSiteSettingApplication.java]: It's a basic JSON writer that filters the fields specified in the "fields" query param. It supports several levels of hierarchy. 

That way clients can reduce the length of the response. Last PR of ahdezma in liferay.

**Pull request closed without merging**